// ControlH3/lib/schedule/schedule.js
var ScheduleSchema = require('../models/ScheduleSchema');

function schedule (){

}

schedule.prototype.addSchedule = function(schedule, callback) {
	var scheduleSchema = new ScheduleSchema({
		idUser : schedule.idUser,
		type : schedule.type,
		year : schedule.year, 
		month : schedule.month,
		input : schedule.input,
		finalised : false
	});

	scheduleSchema.save(function(error,result){
		if(error)
    	{
    		callback(error,null);
    	}
    	else
    	{
    		if(!result){
    			callback(null,null);		
    		}
    		else
    		{
    			callback(null,result);	
    		}
    	}
	});
};

schedule.prototype.getNoFinalizedschedule = function(idUser, callback) {
	ScheduleSchema.findOne({idUser : idUser, finalised : false},function(error, scheduleResult)
  	{
	    if (error){
			  callback(err,null);
	    }
	    else{
	    	if(!scheduleResult){
	   			callback(null,null);		
	   		}
	   		else{
	   			callback(null,scheduleResult);
	   		}
	   	}
	});
};

schedule.prototype.finalisedSchedule = function(schedule, callback) {
	ScheduleSchema.findOne({idUser : schedule.idUser, finalised : false},function(error, scheduleResult)
  	{
	    if (error){
			  callback(error,null);
	    }else{
	    	if(!scheduleResult){
	   			callback(null,null);		
	   		}
	   		else{
	   			scheduleResult.output = schedule.output;
	   			scheduleResult.labored = schedule.labored;
	   			scheduleResult.finalised = true;
	   	
	   			scheduleResult.save(function(error, scheduleResult){
	   				if (error){
						return callback(error,null);
				    }
				    else{
				    	return callback(null,scheduleResult);
				    }	
	   			});
	   		}
	    }
	});
};

schedule.prototype.getHoursByQuery = function(qury, day, callback) {
	ScheduleSchema.find(qury,function(error, scheduleResult)
  	{
  		day = (day != undefined)?day:0;
  		
  		//codigo temporal se debe discutir con Diego Castron
  		if(day == null){
  			day = 0;
  		}

  		if (error){
			  callback(error,null);
	    }
	    else{
	    	if(!scheduleResult){
	   			callback(null,null);		
	   		}
	   		else{
	   			var officeHours = 0;
	   			var houseHours = 0;
	   			for(hours in scheduleResult){
					var inputHour = new Date(scheduleResult[hours].input); 
					var outHour = new Date(scheduleResult[hours].output); 
	
					if (day <= inputHour.getDate()){
		   				if(scheduleResult[hours].type == "house"){
							houseHours += ((outHour-inputHour)/3600000);
		   				}else if(scheduleResult[hours].type == "office"){
							officeHours += ((outHour-inputHour)/3600000);
		   				}
					}
	   			}
	   			callback(null,{
	   				record : scheduleResult,
	   			 	officeHours : officeHours,
	   			 	houseHours : houseHours,
	   			 	total : officeHours + houseHours
	   			});
	   		}
	   	}
	});
};

/*
	result = JSON.parse(result);
	var fecha= new Date();
	fecha.getSeconds(result.timestamp + 1000 * 3600);
	a = new Date("February 13, 2014 04:29:00");
	b = new Date("February 12, 2014 03:00:00");
	//La diferencia se da en milisegundos así que debes dividir entre 1000
	var c = ((a-b)/3600000);
  	console.log(result.timestamp + 1000 *3600);		
	res.send('Hora fecha	: ' + fecha);
*/
var schedule = module.exports = new schedule();
