// ControlH3/lib/horario/horario.js
var HorarioSchema = require('../models/horarioSchema');

function Horario (){

}

Horario.prototype.addHorario = function(horario, callback) {
	var horarioSchema = new HorarioSchema({
		idUsuer : horario.idUsuer,
		type : horario.type,
		year : horario.year, 
		month : horario.month,
		input : horario.input,
		finalised : false
	});

	horarioSchema.save(function(error,result){
		if(error)
    	{
    		callback(error,null);
    	}
    	else
    	{
    		if(!result){
    			callback(null,null);		
    		}
    		else
    		{
    			callback(null,horarioSchema);	
    		}
    	}
	});
};

Horario.prototype.getNoFinalizedHorario = function(idUsuer, callback) {
	HorarioSchema.findOne({idUsuer : idUsuer, finalised : false},function(error, horarioResult)
  	{
	    if (error) 
	    {
			  callback(err,null);
	    }
	    else
	    {
	    	if(!horarioResult){
	   			callback(null,null);		
	   		}
	   		else
	   		{
	   			callback(null,horarioResult);
	   		}
	   	}
	});
};

Horario.prototype.finalisedHorario = function(horario, callback) {
	HorarioSchema.findOne({idUsuer : horario.idUsuer, finalised : false},function(error, horarioResult)
  	{
	    if (error) 
	    {
			  callback(error,null);
	    }
	    else
	    {
	    	if(!horarioResult){
	   			callback(null,null);		
	   		}
	   		else
	   		{
	   			horarioResult.output = horario.output;
	   			horarioResult.labored = horario.labored;
	   			horarioResult.finalised = true;
	   	
	   			horarioResult.save(function(error, horarioResult){
	   				if (error) 
				    {
						return callback(error,null);
				    }
				    else
				    {
				    	return callback(null,horarioResult);
				    }	
	   			});
	   		}
	    }
	});
};

var Horario = module.exports = new Horario();
