// ControlH3/lib/agenda/agenda.js
var AgendaSchema = require('../models/agendaSchema')

function Agenda()= {

}

Agenda.prototype.addAgenda= function(user,agenda,callback){
	var agendaShema= new AgendaSchema({
		idUser: user,
		vigente: true,
		lunes: agenda.lunes,
		martes: agenda.martes,
		miercoles: agenda.miercoles,
		jueves: agenda.juves,
		viernes: agenda.viernes,
		sabado: agenda.sabado,
		domingo: agenda.domingo
	});
	this.getAgendaVigenteByIdUser(user,function(error,agendaVigente){
		if(error){
			return callback(error,null);
		}
		else{
			if(agendaVigente){
				agendaVigente.vigente= false;
				agendaVigente.save(function(err,agendaVigente){
					return callback(error,null);
				});				
			}
		}
	});
	agendaSchema.save(function(err,agenda){
		if(err){
			return callback(err,null);			
		}
		else{
			if(agenda){
				return callback(null,agenda);
			}
			else{
				return callback(null,null);
			}
		}
	});
};

Agenda.prototype.getAgendaVigenteByIdUser= function(idUser,callback){
	AgendaSchema.findOne({idUser:idUser,vigente:true},function(error,agendaResult){
		if(error){
			callback(error,null);			
		}
		else{
			if(agendaResult){
				callback(null,agendaResult);
			}
			else{
				callback(null,null);
			}
		}
	});
}

Agenda.prototype.getTodasLasAgendasUsuario= function(idUser,callback){
	AgendaSchema.find({idUser:idUser},function(error,agendas){
		if(error){
			callback(error,null);			
		}
		else{
			if(agendaResult){
				callback(null,agendaResult);
			}
			else{
				callback(null,null);
			}
		}
	});
}

Agenda.prototype.getAgendaNoVigenteByIdUser= function(idUser,callback){
	AgendaSchema.findOne({idUser:idUser,vigente:false},function(error,agendaResult){
		if(error){
			callback(error,null);			
		}
		else{
			if(agendaResult){
				callback(null,agendaResult);
			}
			else{
				callback(null,null);
			}
		}
	});
}

Agenda.prototype.getAgendasVigentesUsuarios= function(callback){
	AgendaSchema.findOne({vigente:true},function(error,agendasUsuarios){
		if(error){
			callback(error,null);			
		}
		else{
			if(agendaResult){
				callback(null,agendasUsuarios);
			}
			else{
				callback(null,null);
			}
		}
	});
}

var Agenda= module.exports= new Agenda();
