// ControlH3/lib/requirement/requirement.js
var RequirementSchema = require('../models/requirementSchema')

function Requirement(){

}

Requirement.prototype.addRequirement= function(info,callback){ 
    var data = {
        backlogId : info.backlogId,
        idUser : info.idUser,
        task : info.task,
        detail : info.details,
        coments : info.coments,
        remainingTime : info.remainingTime,
        priority : info.priority,
        status : status
    } 
    
    var requirement= new RequirementSchema(data);
    requirement.save(function(error, response){
        if(error){
            return callback(error,null);            
        }
        else{
            if(!response){
                return callback(null,null);             
            }
            else{
                return callback(null,response);
            }
        }
    });
};

Requirement.prototype.getRequirementsByBacklogId = function(idBacklog, callback){
    RequirementSchema.find({backlogId : idBacklog}, function(error,response){
        if(error){
            return callback(error,null);            
        }
        else{
            if(!response){
                return callback(null,null);             
            }
            else{
                return callback(null,response);
            }
        }
    });
};

Requirement.prototype.getRequirementById = function(id, callback){
    RequirementSchema.findOne({_id : id}, function(error,response){
        if(error){
            return callback(error,null);            
        }
        else{
            if(!response){
                return callback(null,null);             
            }
            else{
                return callback(null,response);
            }
        }
    });
};

Requirement.prototype.patchRequirementById = function(id, info, callback){
    
}

var Requirement= module.exports= new Requirement();
