// ControlH3/lib/ControlH3.js
'use strict';

var User = require('./user/user');
var WorkSpace = require('./workSpace/workSpace');
var Schedule =  require('./schedule/schedule');
var Project = require('./project/project');
var Backlog = require('./backlog/backlog');
var Requirement = require('./requirement/requirement');
var utilities = require('./utilities');


function ControlH3 (){

}

/----------------------------------------------------- User managment and basic services  ------------------------------------------------/

ControlH3.prototype.signIn = function(name, password, date, mac, callback) {
	var user = {
		name : name, 
		password : password
	};

	User.getUserByNameAndPassword(user, function(err, userResult){
		if(err){
			return callback(err, null);
		}
		else
		{
			if (userResult == null) {
				callback("Invalid login name or password", null);
			}else if(!userResult.logged){
				var type = (!mac)?'house':'office';				
				var workSpace = {
					idUser : userResult._id,
					type : type,
					year : date.getFullYear(),
					month : date.getMonth() + 1,
					input : date	
				};

				WorkSpace.addWorkSpace(workSpace, function(err, workSpace){
					if(err){
						return callback(err, null);
					}
					else
					{
						userResult.logged = true;
						userResult.state = type;
						userResult.save();
						return callback(null,workSpace);
					}
				});		
				
			}else{
				return callback("User already logged", null);			
			}
		}	
	});
};


ControlH3.prototype.signOut = function(name, password, date, mac, labored, callback) {
	var user = {
		name : name, 
		password : password
	};

	User.getUserByNameAndPassword(user, function(err, userResult){
		if(err){
			return callback(err, null);
		}
		else{
			if (!userResult) {
				callback("Invalid login name or password", null);
			}else if(userResult.logged){
				var type = (!mac)?'house':'office';
				if(userResult.state == type){					
					var workSpace = {
						idUser : userResult._id,
						output : date,
	   					labored : labored,
	   					finalised : true
					};
					WorkSpace.finalisedWorkSpace(workSpace, function(error, workSpace){
						if(error){
							return callback(error, null);
						}else{
							userResult.logged = false;
							userResult.state = 'out';
							userResult.save();
							return callback(null, workSpace);
						}					
					});
					
				}
				else{
					return callback("marque salida desde la oficina o por internet "
						+"al igual que marco entrada",null);
				}
			}
			else{
				return callback("Ud no se encuentra logged",null);
			}
		}
	});
};

ControlH3.prototype.passwordChange= function(form,callback){
	User.changePassword(form,function(err,result){
		if(err){
			callback(err, null);
		}else{			
			if(!result){
				return callback("User was not found",null);
			}	
			else{
				return callback(null,result);
			}
		}
	});
};

ControlH3.prototype.signUp = function(user, callback) {
	User.addUser(user, function(err, result){
		if(err){
			return callback(err, null);
		}else{
			return callback(null, result);
		}
	});
};

ControlH3.prototype.workingNow = function(callback){
	User.getUsersLoggedOn(function(err, result){
		if(err){
			return callback(err, null);
		}else{
			return callback(null, result);
		}
	});
};

ControlH3.prototype.hi= function(callback){
	console.log("hi, good day!!!");
	utilities.getHora(callback);
};

/--------------------------------------- Schedule managment : (Post, Get by user, Get All, and Patch)  -----------------------------------/ 

ControlH3.prototype.addSchedule= function(schedule,callback){
	var query = {
		_id : schedule.idUser
	};
	User.getUserByQuery(query,function(error,response){
		if(error){
			return callback(error,null);
		}else if (!response){
			return callback("The user to be added the schedule does not exist",null);
		}else{
			Schedule.addSchedule(schedule,function(err,result){
				if(err){
					return callback(err, null);
				}else{
					return callback(null, result);
				}
			});
		}

	});	
};

ControlH3.prototype.getSchedulesByIdUser= function(user,callback){
	var query = {
		_id : user
	};
	User.getUserByQuery(query,function(error,response){
		if(error){
			return callback(error,null);
		}else if (!response){
			return callback("There is no user with the ID provided",null);
		}else{
			Schedule.getSchedulesByIdUser(user,function(err,result){
				if(err){
					return callback(err,null);
				}else if(!result){
					return callback("The user do not have any registered shedule", null);
				}else{
					return callback(null,result);
				}
			});
		}
	});	
};

ControlH3.prototype.getAllSchedules = function(callback){
	Schedule.getSchedulesUsers(function(err,result){
		if(err){
			return callback(err,null);
		}else if(!result){
			return callback("No registered shedules", null);
		}else{
			return callback(null,result);
		}
	});
};

ControlH3.prototype.patchScheduleById = function(id, info, callback){
	Schedule.patchScheduleById(id,info, function(error,result){
		if(error){
			return callback(error, null);
		}else{
			return callback(null, result);
		}
	});
};

/------------------------------------------ Measurement ranges of hour work in different ways ---------------------------------------------/

ControlH3.prototype.getTotalHours = function(info, callback){
	var query = {
		idUser  : info.idUser,
		year  : info.year,
		month  : info.month,
		finalised : true
	}

	WorkSpace.getHoursByQuery(query, info.day, function(err,result){
		if(err){
			return callback(err,null);
		}else{
			return callback(null,result);
		}
	});
};

ControlH3.prototype.getHoursInDateRange = function (info, callback){
	var split_init = info.initDate.split(".");
	var initDateUTC= Date.UTC(split_init[2],split_init[0]-1,split_init[1]);
	var init_date = new Date(initDateUTC);
	init_date.setDate(init_date.getDate());

	var split_final = info.finalDate.split(".");
	var endDateUTC = Date.UTC(split_final[2],split_final[0]-1,split_final[1]);
	var end_date = new Date(endDateUTC);
	end_date.setDate(end_date.getDate()+1);
	
	var query = {		
		input: {
			$gte : init_date			
		},
		output: {
			$lt  : end_date 
		},
		idUser: info.idUser,
		finalised: true
	}

	WorkSpace.getHoursByQuery(query, null, function (err, hours){		
		if(err){
			return callback(err, null);
		}
		else{
			User.getUserByQuery({_id: info.idUser}, function (err, user){
				if(err){
					return callback(err, null);			
				}
				else if(user == null){
					return callback("The user does not exist", null);
				}
				else{
					return callback(null, {user: user.name, horas: hours});
				}
			});
		}
	});
};

/------------------------------------------ Project managment : (Post, Get All , Get by id and Patch)  -----------------------------------/

ControlH3.prototype.addProject = function (info, callback){
	//Add the code to create a backlog product for the new project
	Project.addProject(info,function(error,response){
		if(error){
			return callback(error,null);
		}else {
			Backlog.addBacklog(response._id, "product", {editable : true}, function(err, result){
				//Considerar armar una respuesta con los objetos proyecto y err en la creacion del backlog
				if(err){
					console.log(err);
					return callback(null, response);
				}else{
					return callback(null, response);
				}
			});			
		}
	});
};

ControlH3.prototype.getAllProjects = function(callback){
	Project.getProjects(function(error,response){
		if(error){
			return callback (error,null);
		}else if (! response){
			return callback ("No registered Projects", null);
		}else{
			return callback (null, response);
		}
	});
};

ControlH3.prototype.getProjectById = function(idProject,callback){
	Project.getProjectById(idProject,function(error,response){
		if(error){
			return callback(null,null);
		}else if(!response){
			return callback("The project requested does not exist", null);
		}else{
			return callback(null,response);
		}
	});
};

ControlH3.prototype.patchProjectById = function(idProject, info, callback){
	Project.patchProjectById(idProject,info, function(error,response){
		if(error){
			return callback(error, null);
		}else{
			return callback(null, response);
		}
	});
};

/------------------------------------------ Backlog managment : (Post, Get All , Get by id and Patch)  -----------------------------------/

ControlH3.prototype.addBacklog = function(idProject, type, data, callback){
	var info = {
		projectId : idProject,
		type : type,
		editable : data.editable
	}; 
	//Antes de agregar el backlog verificar que los requerimientos o historias en items existan, y hacer una sublista con los que cumplen
	Backlog.addBacklog(info, function(error, response){
		if(error){
			return callback(error, null);
		}else if(!response){
			return callback(null, null);
		}else{
			return callback(null, response);
		}
	});
};

ControlH3.prototype.getBacklogsByProjectIdAndType = function(idProject, type,  callback){
	Backlog.getBacklogsByProjectIdAndType(idProject, type, function(error, response){
		if(error){
			return callback(error, null);
		}else if(!response || !response.length){
			return callback("No Backlogs was found", null);
		}else{			
			return callback(null, response);
		}
	});
};

/------------------------------------------ Requirement managment : (Post, Get All , Get by id and Patch) --------------------------------------/

ControlH3.prototype.addRequirement = function(data, callback){
	Requirement.addRequirement(data, function(error, response){
		if(error){
			return callback(error, null);
		}else if(!response){
			return callback(null, null);
		}else{			
			return callback(null, response);
		}
	});
};

ControlH3.prototype.getRequirementsByBacklogId = function(backlogID, callback){
	Requirement.getRequirementsBybacklogId(backlogID, function(error, response){
		if(error){
			return callback(error, null);
		}else if(!response){
			return callback(null, null);
		}else{			
			return callback(null, response);
		}
	});
};

ControlH3.prototype.getRequirementById = function(id, callback){
	Requirement.getRequirementsById(id, function(error, response){
		if(error){
			return callback(error, null);
		}else if(!response){
			return callback(null, null);
		}else{			
			return callback(null, response);
		}
	});
};

var controlh3 = module.exports = exports = new ControlH3;
