// ControlH3/lib/user/User.js
var userSchema = require('../models/userSchema');

function User(){

}

User.prototype.getUsersLoggedOn = function(callback){
	userSchema.find({logged : true},function(error, users){
   if(error){
    callback(err,null);
  }
  else{
    if(!users){
     return callback(null,null);		
   }
   else{
     return callback(null,users);	
   }
 }
});
}


User.prototype.getUserByName = function(name, callback){
	userSchema.findOne({name : name},function(error, user){
   if(error){
    return callback(err,null);
  }
  else{
    if(!user){
     return callback(null,null);		
   }
   else{
     return callback(null,user);	
   }
 }
});
}

User.prototype.getUserByQuery = function (query, callback){
  userSchema.findOne(query, function (err, user){
    if(err){
      return callback(err,null);
    }
    else{
      if(!user){
        return callback(null,null);    
      }
      else{
        return callback(null,user);  
      }
    }
  });
};

User.prototype.changeStateByName = function(name, state, callback){
	userSchema.findOne({name : name},function(error, user){
    if (error){
      return callback(error,null);
    }
    else{
    	if(!user){
        return callback(null,null);		
      }
      else{
        if(user.logged == false && state != "out" ){
          user.logged = true;
          user.state = state;
       }
       else if(user.logged == true && state == "out")
       {
         user.logged = false;
         user.state = state;
       }
       user.save();
       return callback(null, user);	
     }
   }
 });
}

User.prototype.addUser = function(User, callback){
  this.getUserByName(User.name, function(error, response){
    if(error){
      return callback(error,null);
    }
    else{
      if(response){
        return callback("User already exist.", null);
      }
      else{
        var pass = encriptarPassword(User);
        var user = new userSchema({
          name : User.name,
          password : pass,
          state : "out",
          logged : false,
          active : true
        });
        user.save(function(err,user){
          if(err){
            return callback(err, null);
          }
          else{
            if(!user){
              return callback(null, false);    
            }
            else{
              return callback(null, user);
            }
          }
        });
      }
    }
  });  
}

User.prototype.getUserByNameAndPassword = function(User, callback){  
  var pass = encriptarPassword(User);
  userSchema.findOne({name : User.name, password : pass}, function(error, user){   
    if (error){
      return callback(error,null);
    }
    else{
      if(user){
        return callback(null,user);  
      }
      else{
        return callback(null,null);
      }
    }
  }); 
}

User.prototype.changePassword= function(UserSubmit,callback){    
  this.getUserByNameAndPassword(UserSubmit,function(error,user){
    if(error){
      return callback(error,null);
    }else{
      if(user){
        user.password= encriptarPassword({name: UserSubmit.name, password: UserSubmit.newPassword});
        user.save(function(err,user){
          if(err){
            return callback(err, null);
          }
          else{
            if(!user){
              return callback(null, null);
            }
            else{
              return callback(null, user);
            }
          }
        });        
      }
      else{
        return callback("Invalid user or password",null);
      }
    }
  });
}

var encriptarPassword = function(user){
  var crypto = require('crypto');
  var hmac = crypto.createHmac('sha1', user.name).update(user.password).digest('hex');
  return hmac;
};

var User = module.exports = new User();
