// ControlH3/lib/workSpace/workSpace.js
var workSpaceSchema = require('../models/workSpaceSchema');

function WorkSpace (){

}

WorkSpace.prototype.addWorkSpace = function(workSpace, callback) {
	var workSpaceRegister = new workSpaceSchema({
		idUser : workSpace.idUser,
		type : workSpace.type,
		year : workSpace.year, 
		month : workSpace.month,
		input : workSpace.input,
		finalised : false
	});

	workSpaceRegister.save(function(error,result){
		if(error){
    		callback(error,null);
    	}
    	else{
    		if(!result){
    			callback(null, null);		
    		}
    		else{
    			callback(null,result);	
    		}
    	}
	});
};

WorkSpace.prototype.getNoFinalizedWorkSpace = function(idUser, callback) {
	workSpaceSchema.findOne({idUser : idUser, finalised : false},function(error, workSpaceResult){
	    if (error){
			  callback(err,null);
	    }
	    else{
	    	if(!workSpaceResult){
	   			callback(null,null);		
	   		}
	   		else{
	   			callback(null,workSpaceResult);
	   		}
	   	}
	});
};

WorkSpace.prototype.finalisedWorkSpace = function(workSpace, callback) {
	workSpaceSchema.findOne({idUser : workSpace.idUser, finalised : false},function(error, workSpaceResult)
  	{
	    if (error){
			  callback(error,null);
	    }else{
	    	if(!workSpaceResult){
	   			callback(null,null);		
	   		}
	   		else{
	   			workSpaceResult.output = workSpace.output;
	   			workSpaceResult.labored = workSpace.labored;
	   			workSpaceResult.finalised = true;
	   	
	   			workSpaceResult.save(function(error, workSpaceResult){
	   				if (error){
						return callback(error,null);
				    }
				    else{
				    	return callback(null,workSpaceResult);
				    }	
	   			});
	   		}
	    }
	});
};

WorkSpace.prototype.getHoursByQuery = function(qury, day, callback) {
	workSpaceSchema.find(qury,function(error, workSpaceResult)
  	{
  		day = (day != undefined)?day:0;
  		
  		//codigo temporal se debe discutir con Diego Castron
  		if(day == null){
  			day = 0;
  		}

  		if (error){
			  callback(error,null);
	    }
	    else{
	    	if(!workSpaceResult){
	   			callback(null,null);		
	   		}
	   		else{
	   			var register = [];
	   			var officeHours = 0;
	   			var houseHours = 0;
	   			for(hours in workSpaceResult){
					var inputHour = new Date(workSpaceResult[hours].input); 
					var outHour = new Date(workSpaceResult[hours].output); 

					register.push({
						workedFrom : workSpaceResult[hours].type,
						//Dates are sent in a string to be re-interpreted on the client side with the corresponding time zone 
						input : inputHour.toUTCString(),
						output : outHour.toUTCString(),
						labored: workSpaceResult[hours].labored	
					});
	
					if (day <= inputHour.getDate()){
		   				if(workSpaceResult[hours].type == "house"){
							houseHours += ((outHour-inputHour)/3600000);
		   				}else if(workSpaceResult[hours].type == "office"){
							officeHours += ((outHour-inputHour)/3600000);
		   				}
					}
	   			}
	   			callback(null,{
	   				record : register,
	   			 	officeHours : officeHours,
	   			 	houseHours : houseHours,
	   			 	total : officeHours + houseHours
	   			});
	   		}
	   	}
	});
};

/*
	result = JSON.parse(result);
	var fecha= new Date();
	fecha.getSeconds(result.timestamp + 1000 * 3600);
	a = new Date("February 13, 2014 04:29:00");
	b = new Date("February 12, 2014 03:00:00");
	//La diferencia se da en milisegundos así que debes dividir entre 1000
	var c = ((a-b)/3600000);
  	console.log(result.timestamp + 1000 *3600);		
	res.send('Hora fecha	: ' + fecha);
*/
var WorkSpace = module.exports = new WorkSpace();
