// ControlH3/lib/models/agendaSchema.js
var mongoose = require('mongoose'),
	Schema = mongoose.Schema;

var agendaSchema = new Schema({
	idUser : {type : Number, required: true, index: { unique: true}},
	vigente: {type: Boolean, required: true},
	lunes: [{
	 	type : String, required: true, enum: ['house','office'],
		input : {type : Date, required: true},
		output : {type : Date}
	}],
	martes: [{
	 	type : String, required: true, enum: ['house','office'],
		input : {type : Date, required: true},
		output : {type : Date}
	}],
	miercoles: [{
	 	type : String, required: true, enum: ['house','office'],
		input : {type : Date, required: true},
		output : {type : Date}
	}],
	jueves: [{
	 	type : String, required: true, enum: ['house','office'],
		input : {type : Date, required: true},
		output : {type : Date}
	}],
	viernes: [{
	 	type : String, required: true, enum: ['house','office'],
		input : {type : Date, required: true},
		output : {type : Date}
	}],
	sabado: [{
	 	type : String, required: true, enum: ['house','office'],
		input : {type : Date, required: true},
		output : {type : Date}
	}],
	domingo: [{
	 	type : String, required: true, enum: ['house','office'],
		input : {type : Date, required: true},
		output : {type : Date}
	}]
});

module.exports = mongoose.model('agenda', agendaSchema);
