// ControlH3/lib/models/requirementShema.js
var autoIncrement = require('mongoose-auto-increment');
var mongoose = require('mongoose'),
    Schema = mongoose.Schema,
    ObjectId = Schema.ObjectId;

var connection = mongoose.createConnection("mongodb://localhost/ControlH3");
    autoIncrement.initialize(connection);

var requirementSchema= new Schema({
    _id : {type: Number, index: {unique: true}},
    task : {type: String, required: true},
    details :{type : String},
    priority: {type : Number, required: true},
    responsible : {type: Number, required: true},
    //A list of commets added for the devolopers/testers about requirement
    comments : [String],
    remainingTime: {type: Number, required: true},
    priority: {type : String, required: false, enum: ['Very low', 'low', 'Mid','High', 'Urgent']},
    status : {type: String, required: true, enum: ['in Backlog', 'in Sprint', 'in Test', 'tested']};
});

requirementSchema.plugin(autoIncrement.plugin, 'requirement');
module.exports = mongoose.model('requirement', requirementSchema);
