// ControlH3/lib/models/requirementSchema.js
var autoIncrement = require('mongoose-auto-increment');
var mongoose = require('mongoose'),
    Schema = mongoose.Schema,
    ObjectId = Schema.ObjectId;

var connection = mongoose.createConnection("mongodb://localhost/ControlH3");
    autoIncrement.initialize(connection);

var requirementSchema= new Schema({
    _id : {type: Number, index: {unique: true}},    
    backlogId : {type: Number, required : true, index : {unique : false}},
    idUser : {type: Number, required: true, index: {unique: false}},
    task : {type: String, required: true},
    details :{type : String},    
    //A list of commets added for the devolopers/testers about requirement
    comments : [String],
    remainingTime: {type: Number, required: true},
    priority: {type : String, required: true, enum: ['Very low', 'Low', 'Medium','High', 'Urgent']},
    status : {type: String, required: true, enum: ['Discarded','Reassigned','In Backlog', 'In Sprint', 'In Test', 'Tested', 'Completed']}
});

requirementSchema.plugin(autoIncrement.plugin, 'requirement');
module.exports = mongoose.model('requirement', requirementSchema);
